eip-0014.md

```diff
@@ -0,0 +1,877 @@
+            else validRedemption
+
+    sigmaProp(
+        validSuccessorScript &&
+        preservedFeeConfig &&
+        preservedPoolNFT &&
+        validLP &&
+        validY &&
+        noMoreTokens &&
+        validAction && 
+        validStorageRent
+    )
+}
+```
+
+#### Swap proxy-contract (N2T)
+
+Swap contract ensures a swap is executed fairly from a user's perspective. The contract checks that:
+* Assets are swapped at actual price derived from pool reserves. `X_output = X_reserved * Y_input * fee_num / (Y_reserved * 1000 + Y_input * fee_num)`
+* Fair amount of DEX fee held in ERGs. `F = X_output * F_per_token`
+* A minimal amount of quote asset received as an output in order to prevent front-running attacks.
+
+Once published swap contracts are tracked and executed by ErgoDEX bots automatically.
+Until a swap is executed, it can be cancelled by a user who created it by simply spending the swap UTXO.
+
+##### Sell Ergs
+###### Contract parameters
+Constant            | Type       | Description
+--------------------|------------|---------------
+Pk                  | SigmaProp  | User PublicKey
+FeeNum              | Long       | Pool fee numerator (must be taken from pool params)
+QuoteId             | Coll[Byte] | Quote asset ID. This is the asset we are buying from the pool
+MinQuoteAmount      | Long       | Minimal amount of quote asset
+BaseAmount          | Long       | The amount of nanoErgs to sell
+DexFeePerTokenNum   | Long       | Numerator of the DEX fee in nanoERGs per one unit of quote asset
+DexFeePerTokenDenom | Long       | Denominator of the DEX fee in nanoERGs per one unit of quote asset
+PoolNFT             | Coll[Byte] | ID of the pool NFT (Used as a reference to a concrete unique pool)
+
+```scala
+{   // contract to sell Ergs and buy Token
+    val PoolNFT = fromBase64("ERERERERERERERERERERERERERERERERERERERERERE=") // TODO replace with actual NFT ID
+    val QuoteId = fromBase64("ERERERERERERERERERERERERERERERERERERERERERE=") // TODO replace with actual Quote token id
+    val DexFeePerTokenNum = 10L // TODO replace with actual Dex fee num
+    val DexFeePerTokenDenom = 1000L // TODO replace with actual Dex fee denom
+    val Pk = sigmaProp(true)  // TODO replace with actual sigma prop
+    val FeeNum = 997L // TODO replace with actual feeNum (997 corresponds to 0.3% fee)
+    val BaseAmount = 100L // TODO replace with actual base amount
+    val MinQuoteAmount = 20 // TODO replace with actual minimum quote amount
+
+
+    val validTrade =
+        if (INPUTS.size == 2) { // we don't permit more than one swap per tx
+            val FeeDenom = 1000
+
+            val poolIn  = INPUTS(0)
+            val poolNFT = poolIn.tokens(0)._1
+
+            val poolY_token   = poolIn.tokens(2)
+            val poolY_tokenId = poolY_token._1
+
+            val poolReservesX = poolIn.value
+            val poolReservesY = poolY_token._2
+            val validPoolIn   = poolNFT == PoolNFT && poolY_tokenId == QuoteId
```

Comment:
It seems `poolY_tokenId == QuoteId` check is excessive. Can be checked off chain when pool is selected.

Reply:
This check is discarded currently.